metronome/checkpointing/models/props/src/io/iohk/metronome/checkpointing/models/CheckpointCertificateProps.scala

```diff
@@ -0,0 +1,204 @@
+package io.iohk.metronome.checkpointing.models
+
+import cats.data.NonEmptyList
+import io.iohk.metronome.checkpointing.CheckpointingAgreement
+import io.iohk.metronome.crypto.ECKeyPair
+import io.iohk.ethereum.rlp
+import io.iohk.metronome.hotstuff.consensus.{
+  LeaderSelection,
+  Federation,
+  ViewNumber
+}
+import io.iohk.metronome.hotstuff.consensus.basic.{
+  Secp256k1Signing,
+  Phase,
+  QuorumCertificate
+}
+import java.security.SecureRandom
+import org.scalacheck._
+import org.scalacheck.Prop.{forAll, forAllNoShrink, propBoolean}
+import scodec.bits.ByteVector
+import scala.util.Random
+
+object CheckpointCertificateProps extends Properties("CheckpointCertificate") {
+  import ArbitraryInstances._
+  import Arbitrary.arbitrary
+  import RLPCodecs._
+
+  // Testing with real signatures is rather slow.
+  override def overrideParameters(p: Test.Parameters): Test.Parameters =
+    p.withMinSuccessfulTests(25)
+
+  implicit val genFederation: Gen[
+    (
+        Federation[CheckpointingAgreement.PKey],
+        Vector[CheckpointingAgreement.SKey]
+    )
+  ] =
+    for {
+      seed <- arbitrary[Array[Byte]]
+      rnd = new SecureRandom(seed)
+      fedSize <- Gen.choose(3, 10)
+      keys = Vector.fill(fedSize)(ECKeyPair.generate(rnd))
+      fed  = Federation(keys.map(_.pub))(LeaderSelection.RoundRobin)
+    } yield fed.fold(sys.error, identity) -> keys.map(_.prv)
+
+  implicit val arbFederation =
+    Arbitrary(genFederation.map(_._1))
+
+  implicit val signing =
+    new Secp256k1Signing[CheckpointingAgreement](
+      (phase, viewNumber, blockHash) =>
+        List(
+          rlp.encode(phase),
+          rlp.encode(viewNumber),
+          rlp.encode(blockHash)
+        ).map(ByteVector(_)).reduce(_ ++ _)
+    )
+
+  def buildChain(
+      n: Int,
+      parentLedger: Ledger,
+      ancestors: NonEmptyList[Block.Header]
+  ): Gen[NonEmptyList[Block.Header]] =
+    if (n == 0) {
+      Gen.const(ancestors.reverse)
```

Comment:
It is unclear for me why we reverse as the last step. Maybe a short description what this Gen achieve?

Reply:
Added docstring.